### PR TITLE
Fix problem

### PR DESCRIPTION
**Pull request by geniusgordon**
delete testcase confirm
change judge type will remove file confirm
hide problem tags while contest
check user permission before downloading testcase and special, parital judge code

### CODE REVIEW

File: problem/static/problem/js/edit.js [hunk 1]
```diff
@@ -108,6 +110,13 @@ $(document).ready(function() {
         });
         return false;
     });
+    $("#problem_info").submit(function() {
+        if (judge_type == 'LOCAL_SPECIAL' || judge_type == 'LOCAL_PARTIAL') {
+            if (judge_type != $("#id_judge_type").val())
+            return confirm("You've changed the judge type\nThis will remove the original codes\nAre you sure?");
```
Comment (nickelcheng) [inline problem/static/problem/js/edit.js]: This line is too long.

File: problem/templates/problem/problem.html [hunk 2]
```diff
@@ -92,28 +92,30 @@ <h4 class='modal-title'>Limits</h4>
...
+        {% endfor %}
+      {% endif %}
+    </div>
+    {% if tag_form and user.is_authenticated %}
+      <form id='addTag'>
+        {% csrf_token %}
+        <div class='input-group'>
+          {{ tag_form.tag_name }}
+          <span class="input-group-btn">
+            <button id='addTagButton' class="btn btn-default" type="button">Add Tag</button>
```
Comment (nickelcheng) [inline problem/templates/problem/problem.html]: This line is too long.

File: problem/views.py [hunk 3]
```diff
@@ -123,8 +123,14 @@ def edit(request, pid=None):
                 with open('%s%s.h' % (PARTIAL_PATH, problem.pk), 'w') as t_in:
                     for chunk in request.FILES['partial_judge_header'].chunks():
                         t_in.write(chunk)
+            problem = verify_problem_code(problem)
+            if problem.has_special_judge_code and problem.judge_type != problem.SPECIAL:
```
Comment (nickelcheng) [inline problem/views.py]: This line is too long

File: problem/views.py [hunk 4]
```diff
@@ -123,8 +123,14 @@ def edit(request, pid=None):
...
+                os.remove('%s%s%s' % (SPECIAL_PATH, problem.pk, file_ex))
+            if problem.has_partial_judge_code and problem.judge_type != problem.PARTIAL:
```
Comment (nickelcheng) [inline problem/views.py]:
```
if problem.judge_type != problem.PARTIAL:
    if problem.has_partial_judge_code:
        os.remove('%s%s%s' % (PARTIAL_PATH, problem.pk, file_ex))
    if problem.has_partial_judge_header:
        os.remove('%s%s.h' % (PARTIAL_PATH, problem.pk))
```

**Comment by geniusgordon** [comment]
fixed

**Comment by nickelcheng** [comment]
+1, @CHIAHSINHOU please test

**Comment by nickelcheng** [comment]
got a syntax error
screenshot from 2015-05-27 11 58 36

**Comment by geniusgordon** [comment]
fixed QAQ

**Comment by nickelcheng** [comment]
download partial and download special also need to do permission checking.

File: problem/views.py [hunk 5]
```diff
@@ -264,9 +272,19 @@ def preview(request):
     return render_index(request, 'problem/preview.html', {'problem': problem, 'preview': True})
 
 def download_testcase(request, filename):
+    user = validate_user(request.user)
```
Comment (nickelcheng) [inline problem/views.py]: I suggest to use ```@login_required()```

**Comment by geniusgordon** [comment]
fixed

File: problem/templates/problem/problem.html [hunk 6]
```diff
@@ -92,28 +88,32 @@ <h4 class='modal-title'>Limits</h4>
     <h3>Partial Judge Header</h3>
     <a href="{% url 'problem:download_partial' problem.headername %}">{{ problem.headername }}</a>
   {% endif %}
+  {% if not problem.in_contest %}
```
Comment (nickelcheng) [inline problem/templates/problem/problem.html]: I think it's better to disable ALL problem's tag.

Comment (nickelcheng) [inline problem/templates/problem/problem.html]: I mean add new tag function

File: problem/views.py [hunk 7]
```diff
@@ -263,16 +271,37 @@ def preview(request):
...
     return response
 
 def download_partial(request, filename):
+    user = validate_user(request.user)
+    pid = filename.split('.')[0]
+    try:
+        problem = Problem.objects.get(pk=pid)
+    except: 
+        raise Http404()
+    if not has_problem_auth(user, problem):
```
Comment (nickelcheng) [inline problem/views.py]: Is partial judge code designed to be available for everyone?
@CHIAHSINHOU?

File: problem/views.py [hunk 8]
```diff
@@ -263,16 +271,37 @@ def preview(request):
...
         f = open(TESTCASE_PATH+filename, "r")
     except IOError:
+        logger.warning("open testcase %s error" % filename)
         raise Http404()
     response = HttpResponse(FileWrapper(f), content_type="text/plain")
     response['Content-Disposition'] = 'attachment; filename=' + filename
     return response
 
 def download_partial(request, filename):
+    user = validate_user(request.user)
```
Comment (nickelcheng) [inline problem/views.py]: I prefer to use ```@login_required()```

Comment (geniusgordon) [inline problem/views.py]: fixed

File: problem/static/problem/js/edit.js [hunk 9]
```diff
@@ -96,6 +96,8 @@ $(document).ready(function() {
         return false;
     });
     $("body").on('click', '.del_testcase_btn', function(e) {
```
Comment (nickelcheng) [inline problem/static/problem/js/edit.js]: I'm wondering why this function never returns true?

**Comment by CHIAHSINHOU** [comment]
I can still read the tags while a contest is running.

**Comment by nickelcheng** [comment]
for problems not in contest, user can see their tag but can not add

**Comment by CHIAHSINHOU** [comment]
No, I meant I could still read the tags even though the problem was in a running contest.

**Comment by geniusgordon** [comment]
fixed

**Comment by nickelcheng** [comment]
@CHIAHSINHOU please test again,
@geniusgordon please merge the latest code.

**Comment by CHIAHSINHOU** [comment]
This PR is ok for me.

**Comment by geniusgordon** [comment]
merged

**Comment by bruce3557** [comment]
+1, going to merge

**Pull request merged**